File: python/pymess/equation.py
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
#               2009-2018
#

""" The equation class is the base class for all equation objects
    used in Py-M.E.S.S.. Each equation, Lyapunov or Riccati, needs to be derived
    from this class to ensure that the operators are properly defined."""

class Equation(object):
    r""" The equation class is the base class for all equation objects
    used in MESS. Each equation, Lyapunov or Riccati, needs to be derived
    from this class to ensure that the operators are properly defined."""
    b = None
    c = None
    k = None
    rhs = None
    name = ""
    dim = 0
    options = None

    def __init__(self, name="", opt=None, dim=0):
        r""" Constructor of the object. It only sets the name and the options
        to internal variables.  It should be called from the derived class
        using the super mechanism in order to include future changes directly. """
        self.name = name
        self.options = opt
        self.dim = dim

    def __repr__(self):
        r""" Return the representation string. """
        return "Equation: " + self.name

    def __str__(self):
        r""" Return the object as string """
        return "Equation: " + self.name

    # Operator A
    def ax_generate(self):
        r"""Prepares the application of :math:`A` to a right hand side vector or
        matrix. It is called onces before :math:`x = Ay` is called the first time.
        """
        pass

    def ax_clear(self):
        r"""Finalizes the application of the :math:`A` operator. It is called
        after :math:`x = Ay` is applied the last time."""
        pass

    def ax_apply(self, op, y):
        r"""Applies :math:`A` to a right hand side. It has to return
        the result of :math:`x = Ay`. """
        raise NotImplementedError("A_apply needs to be implemented")

    # Operator E
    def ex_generate(self):
        r"""Prepares the application of :math:`E` to a right hand side vector or
        matrix. It is called onces before :math:`x = Ey` is called the first time.
        """
        pass

    def ex_clear(self):
        r"""Finalizes the application of the :math:`E` operator. It is called
        after :math:`x = Ey` is applied the last time."""
        pass

    def ex_apply(self, op, y):
        r"""Applies :math:`E` to a right hand side. It has to return
        the result of :math:`x = Ey`."""
        raise NotImplementedError("ex_apply needs to be implemented")

    # Operator inv(A)
    def ainv_generate(self):
        r"""Prepares the application of the inverse of the :math:`A` operator on a
        given right hand side vector or matrix. It is called onces before :math:`Ax = y` is called the first time.
        """
        pass

    def ainv_clear(self):
        r"""Finalizes the application of the inverse of the :math:`A` operator. It is called
        after :math:`Ax = y` is used the last time."""
        pass

    def ainv_apply(self, op, y):
        r"""Applies the inverse of :math:`A` to a right hand side. It has to return
        the solution of :math:`Ax =y`. """
        raise NotImplementedError("ainv_apply needs to be implemented")

    # Operator inv(E)
    def einv_generate(self):
        r"""Prepares the application of the inverse of the :math:`E` operator on a
        given right hand side vector or matrix. It is called onces before :math:`Ex = y` is called the first time.
        """
        pass

    def einv_clear(self):
        r"""Finalizes the application of the inverse of the :math:`E` operator. It is called
        after :math:`Ex=y` is used the last time."""
        pass

    def einv_apply(self, op, y):
        r"""Applies the inverse of :math:`E` to a right hand side. It has to return
        the solution of :math:`Ex = y`."""
        raise NotImplementedError("einv_apply needs to be implemented")

    # Operator ApE
    def apex_generate(self, p):
        r"""Prepares the application of the :math:`A+pE` operator.
        It is called onces before apex_apply is called the first time."""
        pass

    def apex_clear(self):
        r""" Finalizes the application of the inverse of the :math:`A+pE` operator. It is called
        after apex_apply is used the last time."""
        pass

    def apex_apply(self, op, p, idx_p, y):
        r""" Applies function the operator of :math:`A+pE`. It has to return :math:`x=(A+pE)y`."""
        raise NotImplementedError("apex_apply needs to be implemented")

    # Operator ApE
    def apeinv_generate(self, p):
        r"""Prepares the application of the :math:`(A+pE)^{-1}` operator.
        It is called onces before apeinv_apply is called the first time."""
        pass

    def apeinv_clear(self):
        r""" Finalizes the application of the :math:`(A+pE)^{-1}` operator. It is called
        after apex_apply is used the last time."""
        pass

    def apeinv_apply(self, op, p, idx_p, y):
        r""" Applies the inverse of :math:`A+pE` to a right hand side. It has to return
        the solution of :math:`(A+pE)x = y`."""
        raise NotImplementedError("apeinv_apply needs to be implemented")

    # Additional Helpers
    def parameter(self, arp_p, arp_m, b=None, k=None):
        r""" The parmeter function has to return the shift parameters. If None
        is returned shift paramter will be automatically determined.
        The Shift parameter strategy is determined by the options structure.
        """
        raise NotImplementedError("parameter needs to be implemented")

    def _test(self, t, s):
        return "{0:s} - {1:d} - {2:d}".format(self.name, t, s)

    def __delattr__(self, key):
        raise TypeError(
            "Trying to delete attributes of %r is not allowed" % self)

File: matlab/mexmess/equations/equation.m
% EQUATION Class represents an abstract interface for user defined equations.
%
%   EQUATION represents an abstract interface for user defined equations
%   using callback functionality.
%
%   EQUATION properties:
%       dim         - int, state space dimension.
%       eqn_type    - Type of Equation.
%       B           - Control to state matrix.
%       C           - State to output matrix.
%       RHS         - Matrix, right hand side of equation.
%
%   EQUATION methods:
%
%       AX_generate     -   Prepares the application of A to a right hand side vector or
%                           matrix. It is called onces before x = Ay is called the first time.
%       AX_apply        -   Applies A to a right hand side. It has to return the result of y = Ax.
%       AX_clear        -   Finalizes the application of the A operator. It is called after y = Ax is applied the last time.
%
%       EX_generate     -   Prepares the application of E to a right hand side vector or
%                           matrix. It is called onces before x = Ey is called the first time.
%       EX_apply        -   Applies E to a right hand side. It has to return the result of y = Ex.
%       EX_clear        -   Finalizes the application of the E operator. It is called after y = Ex is applied the last time.
%
%       AINV_generate   -   Prepares the application of the inverse of the inv(A) operator on a
%                           given right hand side vector or matrix. It is called onces before y = inv(A)x is called the first time.
%       AINV_apply      -   Applies the inverse of A to a right hand side. It has to return the solution of Ax = y.
%       AINV_clear      -   Finalizes the application of the inv(A) operator. It is called after y = inv(A)x is applied the last time.
%
%       EINV_generate   -   Prepares the application of the inverse of the E operator on a
%                           given right hand side vector or matrix. It is called onces before y = inv(E)x is called the first time.
%       EINV_apply      -   Applies the inverse of E to a right hand side. It has to return the solution of Ex = y.
%       EINV_clear      -   Finalizes the application of the inv(E) operator. It is called after y = inv(E)x is applied the last time.
%
%       ApEX_generate   -   Prepares the application of the A+pE operator. It is called onces before ApEX_apply is called the first time.
%       ApEX_apply      -   Applies function the operator of A+pE. It has to return y=(A+pE)x.
%       ApEX_clear      -   Finalizes the application of the inverse of the A+pE operator. It is called after ApEX_apply is used the last time.
%
%       ApEINV_generate -   Prepares the application of the inv(A+pE) operator.
%                           It is called onces before ApEINV_apply is called the first time.
%       ApEINV_apply    -   Applies the inverse of inv(A+pE) to a right hand side. It has to return the solution of y = inv(A+pE)x.
%       ApEINV_clear    -   Finalizes the application of the inverse of the inv(A+pE) operator. It is called after ApEINV_apply is used the last time.
%
%       parameter       -   The parmeter function has to return the shift parameters. If [] is returned shift paramter will be automatically determined.
%                           The Shift parameter strategy is determined by the options structure.
%   Examples / Tests:
%       <a href="matlab:help my_equation">my_equation</a>, <a href="matlab:edit my_equation">my_equation.m</a>
%       <a href="matlab:help test_callback_lyapunov">test_callback_lyapunov</a>, <a href="matlab:edit test_callback_lyapunov">test_callback_lyapunov.m</a>
%       <a href="matlab:help test_callback_riccati">test_callback_riccati</a>, <a href="matlab:edit test_callback_riccati">test_callback_riccati.m</a>
%
%   See also MESS_EQUATION_T, MESS_LRADI, MESS_LRNM, MESS_OPTIONS, MYEQUATION, TEST_CALLBACK_LYAPUNOV, TEST_CALLBACK_RICCATI.
%
%   Author: Maximilian Behr (MPI Magdeburg, CSC)

%
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program; if not, see <http://www.gnu.org/licenses/>.
%
% Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
%               2009-2018
%


classdef (Abstract) equation < base_equation

    properties(Abstract)
        dim         % int, state space dimension
        eqn_type    % Type of Equation
        B           % control to state matrix
        C           % state to output matrix
        RHS         % matrix, right hand side of equation
    end

    methods(Abstract)
        AX_generate(obj,opt)
        AX_apply(obj,opt,op,x)
        AX_clear(obj)

        EX_generate(obj,opt)
        EX_apply(obj,opt,op,x)
        EX_clear(obj)

        AINV_generate(obj,opt)
        AINV_apply(obj,opt,op,x)
        AINV_clear(obj)

        EINV_generate(obj,opt)
        EINV_apply(obj,opt,op,x)
        EINV_clear(obj)

        ApEX_generate(obj,opt)
        ApEX_apply(obj,opt,op,p, id, x)
        ApEX_clear(obj)

        ApEINV_generate(obj,opt,p)
        ApEINV_apply(obj,opt,op,p,idx,x)
        ApEINV_clear(obj)

        parameter(obj, arp_p, arp_m, B, K)
    end
end
